flask_oidc/__init__.py

```diff
@@ -187,11 +196,115 @@ def init_app(self, app):
         except KeyError:
             pass
 
-    def load_secrets(self, app):
-        # Load client_secrets.json to pre-initialize some configuration
-        return _json_loads(open(app.config['OIDC_CLIENT_SECRETS'],
-                                   'r').read())
-        
+    def init_provider(self, provider):
+        """
+        Do setup for a specific provider
+
+        :param provider: The provider to initialize.
+        :type provider: Dictionary with at lease 'base_url' item
+        """
+
+        secrets = self.load_secrets(provider)
+        assert secrets != None, "Problem with loading secrets"
+
+        self.client_secrets = list(secrets.values())[0]
+        secrets_cache = DummySecretsCache(secrets)
+
+        # Initialize oauth2client
+        self.flow = flow_from_clientsecrets(
+            current_app.config['OIDC_CLIENT_SECRETS'],
+            scope=current_app.config['OIDC_SCOPES'],
+            cache=secrets_cache)
+
+        assert isinstance(self.flow, OAuth2WebServerFlow)
+
+        current_app.config['OIDC_VALID_ISSUERS'] = (self.client_secrets.get('issuer') or GOOGLE_ISSUERS)
+
+    def load_secrets(self, provider):
+
+        try:
+            static_secrets = current_app.config.get('OIDC_CLIENT_SECRETS', None)
+            if static_secrets:
+                return _json_loads(open(static_secrets,'r').read())
+        except Exception as e:
+            raise Exception("Error reading secrets: {}".format(str(e)))
+
+        if not provider:
+            raise Exception("No Provider specified")
+
+        try:
+            url = provider.get('base_url')
+
+            if not url.endswith('/'):
+                url += '/'
+
+            url += ".well-known/openid-configuration"
+
+            logger.debug("Loading: {}".format(url))
+
+            provider_info = json.load(
+                urlopen(url)
+            )
+
+        except Exception as e:
+            raise Exception("Can not obtain well known information: {}".format(str(e)))
+
+        for path in ['issuer', 'registration_endpoint', 'authorization_endpoint', 'token_endpoint', 'userinfo_endpoint', 'jwks_uri']:
+            if path in provider_info and provider_info[path].startswith('/'):
+                provider_info[path] = "{}{}".format(provider.get('base_url'), provider_info[path])
+
+        registration = provider.get('registration', None)
+
+        if not registration:
```

Comment:
So, while I think dynamic discovery is a reasonable idea, fully automatic dynamic registration without any configuration (e.g. `provider['dynamic_registration'] = True`) might be confusing to people.
Specifically because depending on the IdP in use, if people re-register every server start, their previously issued tokens might suddenly become unusable because they can no longer retrieve the token information.

Because of this, I think making the registration explicit would be a good idea.